File: 0004_Median of Two Sorted Arrays.js
/** ***************************
  [Description]
  There are two sorted arrays nums1 and nums2 of size m and n respectively. Find the median of the two sorted arrays. The overall run time complexity should be O(log(m+n)). You may assume nums1 and nums2 cannot be both empty.

  [Input]
  nums1 = [1, 3]
  nums2 = [2, 4]

  [Output]
  2.5

  [Explanation]
  The median is (2 + 3) / 2 = 2.5
*************************** */

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number}
 */
function findMedianSortedArrays(nums1, nums2) {
  const newNums = nums1.concat(nums2);
  newNums.sort((num1, num2) => num1 - num2);
  const totalLength = newNums.length;
  if (totalLength % 2 === 1) {
    return newNums[(totalLength - 1) / 2].toFixed(1);
  }
  return (newNums[totalLength / 2] + newNums[totalLength / 2 - 1]) / 2;
}

function main() {
  const nums1 = [3];
  const nums2 = [-2, -1];
  console.log(findMedianSortedArrays(nums1, nums2));
}

main();

File: python/0004_Median of Two Sorted Arrays.py
# There are two sorted arrays nums1 and nums2 of size m and n respectively.

# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).

# You may assume nums1 and nums2 cannot be both empty.

# Example 1:
# nums1 = [1, 3]
# nums2 = [2]
# The median is 2.0

# Example 2:
# nums1 = [1, 2]
# nums2 = [3, 4]
# The median is (2 + 3)/2 = 2.5

class Solution:
    def findMedianSortedArrays(self, nums1: [int], nums2: [int]) -> float:
        len1, len2 = len(nums1), len(nums2)
        half_length = (len1 + len2) / 2
        index1, index2 = 0, 0
        current, last = None, None
        while index1 + index2 <= half_length:
            last = current
            if index2 >= len2:
                current = nums1[index1]
                index1 += 1
            elif index1 >= len1:
                current = nums2[index2]
                index2 += 1
            elif nums1[index1] < nums2[index2]:
                current = nums1[index1]
                index1 += 1
            else:
                current = nums2[index2]
                index2 += 1

        if (len1 + len2) % 2:  # Odd
            return float(current)
        else:  # Even
            return (current + last) / 2


if __name__ == "__main__":
    s = Solution()
    print(s.findMedianSortedArrays([1], [2, 2, 2, 2]))